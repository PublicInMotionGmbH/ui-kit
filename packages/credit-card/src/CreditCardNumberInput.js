import React from 'react'
import PropTypes from 'prop-types'
import TextMaskInput from 'react-text-mask'

import { TextInput } from '@talixo/text-input'
import { Icon } from '@talixo/icon'

import { buildClassName } from '@talixo/shared'

const moduleName = 'credit-card-number-input'

const propTypes = {
  /** Additional class name passed to wrapper. */
  className: PropTypes.string,

  /** Handler for onBlur event. */
  onBlur: PropTypes.func,

  /** Handler for onChange event. */
  onChange: PropTypes.func,

  /** Handler for onFocus event. */
  onFocus: PropTypes.func,

  /** Value to be displayed inside inputs. */
  value: PropTypes.string
}

const defaultProps = {}

const MASK = [
  // Credit card has 13-19 digits
  /\d/, /\d/, /\d/, /\d/, ' ',
  /\d/, /\d/, /\d/, /\d/, ' ',
  /\d/, /\d/, /\d/, /\d/, ' ',
  /\d/, /\d/, /\d/, /\d/, ' ',
  /\d/, /\d/, /\d/
]

/**
 * Trim value (including \u2000 placeholders).
 *
 * @param {string} value
 * @returns {string}
 */
function trim (value) {
  return value.replace(/[\u2000]+/g, '').trim()
}

/**
 * Component which represents credit card input.
 *
 * @property {object} props
 * @property {string} [props.className]
 * @property {func} [props.onBlur]
 * @property {func} [props.onChange]
 * @property {func} [props.onFocus]
 * @property {string} [props.value]
 *
 * @property {object} state
 * @property {boolean} state.focused
 * @property {string|null} state.value
 *
 * @class
 */
class CreditCardNumberInput extends React.PureComponent {
  state = {
    focused: false,
    value: this.props.value == null ? '' : this.props.value
  }

  /**
   * Update current value in component state,
   * when value is controlled.
   *
   * @param {object} nextProps
   */
  componentWillReceiveProps (nextProps) {
    if (nextProps.value !== this.props.value && nextProps.value != null) {
      this.setState({ value: nextProps.value })
    }
  }

  /**
   * Handle focusing text input.
   *
   * @param {Event|SyntheticEvent} event
   */
  focus = (event) => {
    const { onFocus } = this.props

    // Make sure that user can't click on some place in input, where it guides him
    // TODO: when react-text-mask will properly work with removing placeholder character, remove it
    clearTimeout(this.focusTimeout)
    this.focusTimeout = setTimeout(() => this.setState({
      focused: true
    }))

    if (onFocus) {
      onFocus(event)
    }
  }

  /**
   * Handle losing focus on text input.
   *
   * @param {Event|SyntheticEvent} event
   */
  blur = (event) => {
    const { onBlur } = this.props

    this.setState({ focused: false })

    if (onBlur) {
      onBlur(event)
    }
  }

  /**
   * Handle change on text input.
   *
   * @param {string} value
   */
  change = (value) => {
    const { onChange } = this.props
    value = trim(value)

    if (this.state.value === value) {
      return
    }

    if (this.props.value == null) {
      this.setState({ value })
    }

    if (onChange) {
      onChange(value)
    }
  }

  /**
   * Render credit card input component.
   *
   * @returns {React.Element}
   */
  render () {
    const { className, value, onChange, onFocus, onBlur, ...passedProps } = this.props
    const { focused } = this.state

    // Build class name
    const clsName = buildClassName(moduleName, className)

    return (
      <TextInput
        className={clsName}
        guide={focused}
        keepCharPositions={false}
        mask={MASK}
        placeholderChar={'\u2000'}
        value={this.state.value}
        onChange={this.change}
        onFocus={this.focus}
        onBlur={this.blur}
        left={<Icon name='credit_card' />}
        placeholder='0000 0000 0000 0000'
        InputComponent={TextMaskInput}
        {...passedProps}
      />
    )
  }
}

CreditCardNumberInput.propTypes = propTypes

CreditCardNumberInput.defaultProps = defaultProps

export default CreditCardNumberInput
