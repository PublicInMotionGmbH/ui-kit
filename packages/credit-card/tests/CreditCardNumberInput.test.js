import React from 'react'
import { mount, shallow } from 'enzyme'

import CreditCardNumberInput from '../src/CreditCardNumberInput'

describe('<CreditCardNumberInput />', () => {
  it('renders children correctly', () => {
    const wrapper = shallow(<CreditCardNumberInput />)

    expect(wrapper).toMatchSnapshot()
  })

  it('passes value correctly', () => {
    const value = '12345678910'
    const wrapper = shallow(<CreditCardNumberInput value={value} />)
    const input = wrapper.find('TextInput')

    expect(input.props().value).toEqual(value)
  })

  it('masks correctly', () => {
    const value = '12345678910111213'
    const wrapper = mount(<CreditCardNumberInput value={value} />)
    const input = wrapper.find('input').getDOMNode()
    expect(input.value).toEqual('12345678910111213')
    wrapper.unmount()
  })

  describe('componentWillReceiveProps', () => {
    it('updates current value', () => {
      const wrapper = shallow(<CreditCardNumberInput />)
      wrapper.setProps({ value: '10000' })

      expect(wrapper.state().value).toEqual('10000')
    })

    it('updates current value when value is provided', () => {
      const value = '12345678910'
      const wrapper = shallow(<CreditCardNumberInput value={value} />)
      wrapper.setProps({ value: '10000' })

      expect(wrapper.state().value).toEqual('10000')
    })
  })
})

describe('change', () => {
  let onChange, wrapper, input
  beforeEach(() => {
    onChange = jest.fn()
    wrapper = shallow(<CreditCardNumberInput onChange={onChange} />)

    input = wrapper.find('TextInput')
    input.simulate('change', '2')
  })

  it('onChange is called when input value changes', () => {
    expect(onChange).toHaveBeenCalledTimes(1)
  })

  it('changes state.value to passed value if props.value is null', () => {
    expect(wrapper.state().value).toEqual('2')
  })
})

describe('focus', () => {
  let onFocus, wrapper, input
  beforeEach(() => {
    onFocus = jest.fn()
    wrapper = shallow(<CreditCardNumberInput onFocus={onFocus} />)

    input = wrapper.find('TextInput')
    input.simulate('focus')
  })

  it('onFocus is called when input value changes', () => {
    expect(onFocus).toHaveBeenCalledTimes(1)
  })

  it('changes state.focused to true', () => {
    expect(wrapper.state().focused).toEqual(false)
    setTimeout(() => expect(wrapper.state().focused).toEqual(true))
  })
})

describe('blur', () => {
  let onBlur, wrapper, input
  beforeEach(() => {
    onBlur = jest.fn()
    wrapper = shallow(<CreditCardNumberInput onBlur={onBlur} />)

    input = wrapper.find('TextInput')
    input.simulate('blur')
  })

  it('onBlur is called when input value changes', () => {
    expect(onBlur).toHaveBeenCalledTimes(1)
  })

  it('changes state.focused to false', () => {
    expect(wrapper.state().focused).toEqual(false)
  })
})

describe('trim', () => {
  it('it trims correctly', () => {
    const onChange = jest.fn()
    const wrapper = shallow(<CreditCardNumberInput onChange={onChange} />)

    const input = wrapper.find('TextInput')
    input.simulate('change', '   1234')
    expect(wrapper.state().value).toEqual('1234')
  })
})
