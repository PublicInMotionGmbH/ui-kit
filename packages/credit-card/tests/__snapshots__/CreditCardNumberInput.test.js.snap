// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreditCardNumberInput /> renders children correctly 1`] = `
<TextInput
  InputComponent={[Function]}
  className="talixo-credit-card-number-input"
  error={false}
  guide={false}
  keepCharPositions={false}
  left={
    <Icon
      name="credit_card"
    />
  }
  mask={
    Array [
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
      " ",
      /\\\\d/,
      /\\\\d/,
      /\\\\d/,
    ]
  }
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="0000 0000 0000 0000"
  placeholderChar="â€€"
  type="text"
  value=""
/>
`;
